### DISCUSSION TITLE
Allowing use of `find_batch_file` from user code

### DISCUSSION THREAD

**Discussion by mwichmann** [SCons/scons#4446] [General]
I'm dropping this here because it's kind of hard to call it a "bug", but wanted to ask (ping @jcbrill @bdbaddog ). We don't expose `SCons.Tool.MSCommon.vc.find_batch_file` officially as part of the SCons API. But... one of the larger users of SCons does reach in and use it anyway, and it has changed in SCons 4.6.0.  Which it's absolutely allowed to do.  But it woud be nice to be able to give them some advice.  The usage appears here:

https://github.com/godotengine/godot/blob/1ba920fada9efc8c4476ded50fe673b8db541366/methods.py#L729

I'm not quite sure why they need this... maybe we'll get enlightened over time.  Meanwhile... the new version doesn't take the `env` argument, and does take a `pdir` last argument, which at this point the external project's code doesn't know.  Thoughts on what we could suggest to them?  Or just say "figure it out since you're not using published APIs"?

**Comment by jcbrill** [comment]
The product directory is retrieved once outside of the host/target loop now.

The `env` is only needed for `find_vc_pdir` in case `VSWHERE` is set via the environment.
The `env` is not needed for the `find_batch_file` method once the product directory is known.

Here is a minimal example that *might* work.  Note that both `find_vc_pdir` and `find_batch_file` may raise exceptions.

```
def find_visual_c_batch_file(env):
    from SCons.Tool.MSCommon.vc import (
        get_default_version,
        get_host_target,
        find_vc_pdir,
        find_batch_file,
    )

    # Syntax changed in SCons 4.4.0.
    from SCons import __version__ as scons_raw_version

    scons_ver = env._get_major_minor_revision(scons_raw_version)

    version = get_default_version(env)

    if scons_ver >= (4, 4, 0):
        (host_platform, target_platform, _) = get_host_target(env, version)
    else:
        (host_platform, target_platform, _) = get_host_target(env)
    
    # may raise UnsupportedVersion
    pdir = find_vc_pdir(env, version)
    
    # may raise VisualCException exception(s)
    return find_batch_file(version, host_platform, target_platform, pdir)[0]
```

**EDIT: FIXED VARIABLES IN `find_batch_file` call above**

**This code could likely change dramatically in the future.**

I don't understand why the third argument is ignored for `get_host_target` which is the list of host/target candidate pairs.  The current scons uses this in the inner loop:

```
    ...

    # Find the product directory
    pdir = None
    try:
        pdir = find_vc_pdir(env, version)
    except UnsupportedVersion:
        # Unsupported msvc version (raise MSVCArgumentError?)
        pass
    debug('product directory: version=%s, pdir=%s', version, pdir)

    # Find the host, target, and all candidate (host, target) platform combinations:
    platforms = get_host_target(env, version)
    debug("host_platform %s, target_platform %s host_target_list %s", *platforms)
    host_platform, target_platform, host_target_list = platforms

    d = None
    version_installed = False

    if pdir:

        # Query all candidate sdk (host, target, sdk_pdir) after vc_script pass if necessary
        sdk_queries = []

        for host_arch, target_arch, in host_target_list:
            # Set to current arch.
            env['TARGET_ARCH'] = target_arch
            arg = ''

            # Try to locate a batch file for this host/target platform combo
            try:
                (vc_script, arg, vc_dir, sdk_pdir) = find_batch_file(version, host_arch, target_arch, pdir)
                debug('vc_script:%s vc_script_arg:%s', vc_script, arg)
                version_installed = True
            except VisualCException as e:
                msg = str(e)
                debug('Caught exception while looking for batch file (%s)', msg)
                version_installed = False
                continue
            ...
```
